Fix ArclightCaptures reset before event being handled (#674)

yanang007 opened the pull request:
连锁挖矿模组[Vein Mining](https://www.curseforge.com/minecraft/mc-mods/vein-mining/files/3903527)通过监听BlockBreakEvent事件并在handler中用自己实现的destroyBlock([VeinMiningLogic.java](https://github.com/TheIllusiveC4/VeinMining/blob/439cd78685e2b2ac22d869277a3c2d08689c7b21/forge/src/main/java/top/theillusivec4/veinmining/veinmining/logic/VeinMiningLogic.java#L158-L209))来破坏相连的方块，借此实现兼容其他挖掘方块的事件。

这个pr添加了特别的检查来防止ArclightCaptures捕获的事件上下文被中途触发的后续事件重置。

yanang007 committed:
Fix ArclightCaptures reset before event being handled (IzzelAliz#674)

IzzelAliz commented:
这里实现成一个栈会不会更好？

yanang007 commented:
用栈应该是更合理的实现，按照我的理解，ArclightCaptures在这里应该是负责在destroyBlock这个方法中，把BlockBreakEvent的生命周期延长到destroyBlock操作结束，那么使用栈的话应该也要遵循生命周期来进行入栈出栈操作，在Arclight的实现中他是通过监听BlockBreakEvent来捕获上下文，通过两个地方的注入方法来释放上下文。
```
// ...
int exp = net.minecraftforge.common.ForgeHooks.onBlockBreakEvent(...); // push event context
// ...

// pop event context and forward to Bukkit (in the injected mixin)
return;
```
但是有一个问题是[Vein Mining](https://www.curseforge.com/minecraft/mc-mods/vein-mining/files/3903527)对照原版destroyBlock一比一复刻了一份([VeinMiningLogic.java](https://github.com/TheIllusiveC4/VeinMining/blob/439cd78685e2b2ac22d869277a3c2d08689c7b21/forge/src/main/java/top/theillusivec4/veinmining/veinmining/logic/VeinMiningLogic.java#L158-L209))。这里面没有合适的注入点来插入释放上下文的代码，我之前基于栈的实现就会因此导致入栈的上下文被一直存着。

目前能想到的解决方法包括：
1. 找到一个新的可以兼容的注入点，但是目前没找到一个既不需要大改又行之有效的位置
2. 不管入栈了多少，在reset capture的时候全部出栈，这个大概能解决问题，但是在实现逻辑上会有一点问题
```
// ...
int exp = net.minecraftforge.common.ForgeHooks.onBlockBreakEvent(...); // push event context. stack:[A]
      ->  VeinMiningLogic#harvest(...)
        // ...
        int exp = net.minecraftforge.common.ForgeHooks.onBlockBreakEvent(...); // push derived event context. stack:[A, B]
        // ...
        popResource(...); // stack:[A, B] peek B
        // ...
      <- returned without popping event context
// ...
popResource(...) // populated into wrong context. stack:[A, B] peek B
// pop event context and forward to Bukkit (in the injected mixin). stack:[] -> B, A
return;
```

yanang007 committed:
Fix ArclightCaptures reset before being handled by event stack (IzzelAliz#674)

yanang007 commented:
新的实现将事件分为原生primary和次生secondary，假设mc事件系统目前为单线程：
primary事件代表destroyBlock发出的，可以被Arclight追踪完整生命周期的事件，这部分事件行为不做改变；
secondary事件代表在destroyBlock中onBlockBreakEvent事件响应过程中发出的事件，这部分事件生命周期由模组自行定义，但一定在onBlockBreakEvent函数结束前结束。

因此实现思路是在destroyBlock中，onBlockBreakEvent执行前，通知ArclightCaptures下一个BlockBreakEvent是primary事件，后续发出的BlockBreakEvent均属于secondary事件，使用flag标记，栈会变成[A(p), B(s), C(s), ...(s)]；
onBlockBreakEvent返回后，查找事件栈中的secondary全部响应并转发到bukkit，栈会变成[A(p)] -> B(s), C(s), ...(s)，当前上下文为栈顶上下文，保证为primary事件；
然后继续正常的流程，并在结束时退栈[] -> A(p)。

根据这个流程，原则上来说每次执行destroyBlock都对应到一个primary事件的上下文，其中在destroyBlock外产生的BlockBreakEvent都会被划为secondary事件并在中途被处理掉。

timi137137 commented:
有一股arc api的味道了(x

yanang007 commented:
好耶arc api(√
不过是另一个为forge mod的自由度买单的workaround罢乐

已知的可能的问题是如果destroyBlock在结束前被异常中断了，会导致事件栈中的事件上下文无法被释放，导致某种意义上的内存泄露，可能需要找到上层的方法后面加一个强制清空事件栈的代码？

yanang007 commented on arclight-common/src/main/java/io/izzel/arclight/common/mixin/core/server/management/ServerPlayerGameModeMixin.java:
```diff
@@ -219,11 +220,35 @@ public void handleBlockBreakAction(BlockPos blockPos, ServerboundPlayerActionPac
+        ArclightCaptures.captureNextBlockBreakEventAsPrimaryEvent();
+    }
+
+    @Inject(method = "destroyBlock", at = @At(value = "INVOKE_ASSIGN", target = "Lnet/minecraftforge/common/ForgeHooks;onBlockBreakEvent(Lnet/minecraft/world/level/Level;Lnet/minecraft/world/level/GameType;Lnet/minecraft/server/level/ServerPlayer;Lnet/minecraft/core/BlockPos;)I"))
+    public void arclight$handleSecondaryBlockBreakEvents(BlockPos pos, CallbackInfoReturnable<Boolean> cir) {
+        ArclightCaptures.BlockBreakEventContext breakEventContext = ArclightCaptures.resetSecondaryBlockBreakPlayer();
+        while (breakEventContext != null) {
+            Block block = breakEventContext.getEvent().getBlock();
+            handleBlockBreak(breakEventContext, new BlockPos(block.getX(), block.getY(), block.getZ()));
+            breakEventContext = ArclightCaptures.resetSecondaryBlockBreakPlayer();
+        }
+    }
+
```
这个函数专门处理destroyBlock外创建的BlockBreakEvent，可能需要移动到onBlockBreakEvent末尾？考虑到在非常极端情况下有可能有很野的mod直接越过destroyBlock自己凭空调用onBlockBreakEvent。

yanang007 replied:
这个函数专门处理destroyBlock外创建的BlockBreakEvent，可能需要移动到onBlockBreakEvent末尾？考虑到在非常极端情况下有可能有很野的mod直接越过destroyBlock自己凭空调用onBlockBreakEvent。

yanang007 commented on arclight-common/src/main/java/io/izzel/arclight/common/mixin/core/server/management/ServerPlayerGameModeMixin.java:
```diff
@@ -219,11 +220,35 @@ public void handleBlockBreakAction(BlockPos blockPos, ServerboundPlayerActionPac
         }
     }
 
+    @Inject(method = "destroyBlock", at = @At(value = "INVOKE", target = "Lnet/minecraftforge/common/ForgeHooks;onBlockBreakEvent(Lnet/minecraft/world/level/Level;Lnet/minecraft/world/level/GameType;Lnet/minecraft/server/level/ServerPlayer;Lnet/minecraft/core/BlockPos;)I"))
+    public void arclight$beforePrimaryEventFired(BlockPos pos, CallbackInfoReturnable<Boolean> cir) {
+        ArclightCaptures.captureNextBlockBreakEventAsPrimaryEvent();
+    }
```
这里用于区分是destroyBlock来的事件还是在外面调用的事件，思路是在forge发出事件前，标记下一个待捕获的事件是primary，捕获时就会标记为primary。但是如果存在别的监听器优先级高于Arclight，并引发secondary事件，就会导致primary标记打到错误的事件上。🤔

yanang007 replied:
还有一种想法是在Arclight捕获事件时，去查询调用栈判断是否来自destroyBlock触发，不过感觉查询调用栈好像挺慢的

IzzelAliz replied:
现在就按照兼容vienminer并且不破坏原有mod就行了，别的mod以后遇到再说吧

yanang007 replied:
欧克

yanang007 committed:
Fix ArclightCaptures reset before being handled by event stack (IzzelAliz#674)
- resetBlockBreakPlayer -> popPrimaryBlockBreakEvent
- handleBlockBreak -> handleBlockDrop
- isPrimaryEvent: int -> boolean, as it does not work as intended
- Add warnings
- Add cleaning to avoid memory leak

yanang007 commented:
这一版本重构了函数名，补上了打印警告信息的语句。
增加了强制清空事件栈的保底代码，注入在"tick"和"destroyAndAck"中调用destroyBlock之前，如果没别处调用的话，应该能保证每次进入新的方块破坏事件之前重置ArclightCaptures状态为空，和之前行为一致。
然后vienminer自己触发的新的方块破坏事件对应的物品掉落事件也能正确转发给bukkit了。
